node/src/components/consensus/highway_core/validators.rs

```diff
@@ -209,16 +228,22 @@ impl<T> ValidatorMap<T> {
 
     /// Binary searches this sorted `ValidatorMap` for `x`.
     ///
-    /// If the value is found, `Ok` is returned, containing the index, otherwise `Err`, with the
-    /// first index at which the value is greater than `x`.
-    pub fn binary_search(&self, x: &T) -> Result<ValidatorIndex, ValidatorIndex>
+    /// Returns the lowest index of an entry `>= x`, or `None` if `x` is greater than all entries.
+    pub fn binary_search(&self, x: &T) -> Option<ValidatorIndex>
     where
         T: Ord,
     {
-        self.0
-            .binary_search(x)
-            .map(|i| ValidatorIndex(i as u32))
-            .map_err(|i| ValidatorIndex(i as u32))
+        match self.0.binary_search(x) {
+            Ok(i) => Some(ValidatorIndex(
+                (0..i)
+                    .rev()
+                    .take_while(|j| self.0[*j] == *x)
+                    .last()
+                    .unwrap_or(i) as u32,
+            )),
+            Err(i) if i < self.len() => Some(ValidatorIndex(i as u32)),
```

Comment:
Personally, I find it a bit confusing that for `Ok(i)` case we're using the _value_ of `x` but in the case of `Err(i) if i < self.len()` we're using the indexes. 

It feels like that there are some assumptions about the context in which this method is going to be used baked into this generic implementation of `binary_search`.